File: niro/src/main/java/nl/uva/se/sc/niro/gui/QLHomeController.java
package nl.uva.se.sc.niro.gui;

import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.TextArea;
import javafx.stage.FileChooser;
import javafx.stage.FileChooser.ExtensionFilter;
import javafx.stage.Stage;
import nl.uva.se.sc.niro.parser.QLFormParser;
import org.antlr.v4.runtime.CharStreams;

import java.io.File;
import java.io.IOException;

import static nl.uva.se.sc.niro.model.Ast.QLForm;

public class QLHomeController extends QLBaseController {
    @FXML
    private TextArea errorMessages;

    @FXML
    public void openForm(ActionEvent event) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Select QL form");
        fileChooser.getExtensionFilters().add(new ExtensionFilter("QL Form files", "*.ql"));

        Stage stage = getActiveStage(event);
        File selectedFile = fileChooser.showOpenDialog(stage);
        if (selectedFile != null) {
            try {
                QLForm form = QLFormParser.parse(CharStreams.fromFileName(selectedFile.getAbsolutePath()));
                if (QLFormParser.getParseErrors().isEmpty()) {
                    Scene formScene = createSceneForForm(form);
                    stage.setScene(formScene);
                } else {
                    errorMessages.setText(ErrorUtil.toString(QLFormParser.getParseErrors()));
                    errorMessages.setVisible(true);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private Scene createSceneForForm(QLForm form) throws IOException {
        FXMLLoader loader = new FXMLLoader(getClass().getResource("/nl/uva/se/sc/niro/gui/QLForm.fxml"));
        Parent root = loader.load();
        ((QLFormController) loader.getController()).populateForm(form);
        return new Scene(root);
    }

}

File: niro/src/main/scala/nl/uva/se/sc/niro/gui/controllers/QLHomeController.scala
package nl.uva.se.sc.niro.gui.controllers

import java.io.{ File, IOException }

import javafx.event.ActionEvent
import javafx.fxml.{ FXML, FXMLLoader }
import javafx.scene.control.TextArea
import javafx.scene.{ Parent, Scene }
import javafx.stage.FileChooser
import nl.uva.se.sc.niro.QLFormService
import nl.uva.se.sc.niro.errors.Errors
import nl.uva.se.sc.niro.gui.application.QLForms
import nl.uva.se.sc.niro.model.QLForm

class QLHomeController extends QLBaseController {
  @FXML
  private var errorMessages: TextArea = null

  @FXML
  def openForm(event: ActionEvent): Unit = {
    val fileChooser = new FileChooser
    fileChooser.setTitle("Select QL form")
    fileChooser.getExtensionFilters.add(new FileChooser.ExtensionFilter("QL Form files", "*.ql"))
    val stage = getActiveStage(event)
    val selectedFile: File = fileChooser.showOpenDialog(stage)
    if (selectedFile != null) try {
      val formOrErrors: Either[QLForm, Seq[Errors.Error]] = QLFormService.importQLSpecification(selectedFile)
      formOrErrors match {
        case Left(form) =>
          val formScene = createSceneForForm(form)
          stage.setScene(formScene)
        case Right(errors) =>
          errorMessages.setText(errors.toString)
          errorMessages.setVisible(true)
      }
    } catch {
      case e: IOException =>
        e.printStackTrace()
    }
  }

  @throws[IOException]
  private def createSceneForForm(form: QLForm) = {
    val loader = new FXMLLoader(getClass.getResource(QLForms.FORM_SCREEN))
    val root: Parent = loader.load[Parent]
    loader.getController.asInstanceOf[QLFormController].populateForm(form)
    new Scene(root)
  }
}
